jekyll/_cci2/caching.md

```diff
@@ -236,46 +237,314 @@ The following example demonstrates how to use `restore_cache` and `save_cache` t
+are actually added to classpaths of generated artifacts.
+
+Since Leiningen uses Maven under the hood,
+it has equivalent behavior.
+
+{% raw %}
+
+```yaml
+steps:
+  - restore_cache:
+      keys:
+        # when lock file changes, use increasingly general patterns to restore cache
+        - maven-repo-v1-{{ .Branch }}-{{ checksum "pom.xml" }}
+        - maven-repo-v1-{{ .Branch }}-
+        - maven-repo-v1-
+  - save_cache:
+      paths:
+        - ~/.m2
+      key: maven-repo-v1-{{ .Branch }}-{{ checksum "pom.xml" }}
+```
+
+{% endraw %}
+
+#### npm (Node)
+
+**Safe to Use Partial Cache Restoration?**
+Yes (with NPM5+).
+
+With NPM5+ and a lock file,
+you can safely use partial cache restoration.
+
+{% raw %}
+
+```yaml
+steps:
+  - restore_cache:
+      keys:
+        # when lock file changes, use increasingly general patterns to restore cache
+        - node-v1-{{ .Branch }}-{{ checksum "package-lock.json" }}
+        - node-v1-{{ .Branch }}-
+        - node-v1-
+  - save_cache:
+      paths:
+        - ~/usr/local/lib/node_modules  # location depends on npm version
+      key: node-v1-{{ .Branch }}-{{ checksum "package-lock.json" }}
+```
+
+{% endraw %}
+
+
+#### pip (Python)
+
+**Safe to Use Partial Cache Restoration?**
+Yes (with Pipenv).
+
+Pip can use files
+that are not explicitly specified in `requirements.txt`.
+Using [Pipenv](https://docs.pipenv.org/) will include explicit versioning in a lock file.
+
+{% raw %}
+
+```yaml
+steps:
+  - restore_cache:
+      keys:
+        # when lock file changes, use increasingly general patterns to restore cache
+        - pip-packages-v1-{{ .Branch }}-{{ checksum "Pipfile.lock" }}
+        - pip-packages-v1-{{ .Branch }}-
+        - pip-packages-v1-
+  - save_cache:
+      paths:
+        - ~/.local/share/virtualenvs/venv  # this path depends on where pipenv creates a virtualenv
+      key: pip-packages-v1-{{ .Branch }}-{{ checksum "Pipfile.lock" }}
+```
+
+{% endraw %}
+
+#### Yarn (Node)
+
+**Safe to Use Partial Cache Restoration?**
+Yes.
+
+Yarn has always used a lock file for exactly these reasons.
+
+{% raw %}
+
+```yaml
+steps:
+  - restore_cache:
+      keys:
+        # when lock file changes, use increasingly general patterns to restore cache
+        - yarn-packages-v1-{{ .Branch }}-{{ checksum "yarn.lock" }}
+        - yarn-packages-v1-{{ .Branch }}-
+        - yarn-packages-v1-
+  - save_cache:
+      paths:
+        - ~/.cache/yarn
+      key: yarn-packages-v1-{{ .Branch }}-{{ checksum "yarn.lock" }}
+```
+
+{% endraw %}
 
-Listing multiple keys for restoring a cache increases the odds of a partial cache hit. However, broadening your `restore_cache` scope to a wider history increases the risk of confusing failures. For example, if you have dependencies for Node v6 on an upgrade branch, but your other branches are still on Node v5, a `restore_cache` step that searches other branches might restore incompatible dependencies.
+## Caching Strategy Tradeoffs
```

Comment:
This isn't new — just got carried away with semantic linefeeds because I thought I'd be using copy from this section. 😬